designs/validations.md

```diff
@@ -0,0 +1,288 @@
+		* When using a client to read from the API server, be very careful. Webhooks are supposed to be very fast, remember they are synchronous.
+			* Only run *efficient* queries (avoid full scans) and always think about the number of results you expect to get (depending on the type, the difference can be of orders of magnitude).
+			* If your design performance is not very obvious, benchmark it.
+	* Controllers: any other validations requiring extra information, calls to external systems or simply *too slow* for a webhook.
+
+### Composing validations
+This section lays down the requirements for a framework for validations. Some requisites:
+* Allow to compose and run validations implemented in different places.
+* Make implementing validations as easy as possible.
+* Design mainly for the CLI use-case. Optimize defaults for this scenario and possibly allow custom configuration for others.
+* We want to follow a similar idea to what `cluster.ConfigManager` and other attempts to validation framework propose: a two step process, where validations/defaults can be registered and then run:
+
+```go
+runner := validation.NewRunner()
+runner.Register(myValidation)
+runner.Register(otherValidations...)
+
+err := runner.RunAll(ctx, spec)
+```
+
+#### Validation signature
+Requirements:
+* They might need to contact external systems. In that case they should make use of a `context.Context`.
+* They might need access to the full specification of an EKS-A cluster. In that case, they should use `cluster.Spec`.
+* Validations should be able to, at least, return an error if the input is not valid.
+
+This boils down to take a context and a cluster spec as input and return some type of error as output. We could implement it with functions or an interface, but given that we want to make registering validations as easy as possible and keep the framework API surface as simple as possible, we opt out for a function:
+
+```go
+type Validation func(ctx context.Context, spec *cluster.Spec) error
+```
+
+#### Composability
+We want the framework to allow, not only to register validations implemented in different places, but also validations that are composed of multiple validations. This, although it introduces some complexity, it offers a lot of flexibility to registrars to aggregate validations in different ways and it facilitates reusing existing validations (that tend to be already aggregated in one big method).
+
+##### Aggregating errors
+For validations to be registered as one handler but still be considered as different in the runner result, they should return an `errors.Aggregate`. The framework will understand this and will flatten the result. This allows for multiple levels of nesting.
+
+##### Concurrency
+Most *preflight* validations we currently run in the CLI can be run concurrently as they are not depending on each other. Given most of the time is spent waiting for external calls (mostly HTTP), this could potentially speed up the process.
```

Comment:
```suggestion
Most *preflight* validations we currently run in the CLI can be run concurrently as they are not dependent on each other. For validations that make external calls this can be slow, concurrent invocation of validations offers a convenient speed improvement.
```